`StubbedStderr` in `TestAPI.test_download_submission_autoresume()` breaks `make regenerate-sdk-cassettes`

## Description

The `StubbedStderr` mock hard-codes values that will not correspond to a new run of `make regenerate-sdk-cassettes`, which therefore fails.

https://github.com/freedomofpress/securedrop-client/blob/8d4d9ba9f3ab148c24a216f6572b49170d6793f0/client/tests/sdk/test_api.py#L241-L249


## Steps to Reproduce

```sh-session
$ make regenerate-sdk-cassetttes  # with the Server's "make dev" running
```


## Expected Behavior

Cassettes are regenerated.


## Actual Behavior

```sh-session
________________________________ TestAPI.test_download_submission_autoresume ________________________________

self = <utils.VCRAPI object at 0x7b23c82182d0>, method = 'GET'
path_query = 'api/v1/sources/5b7d9dcf-df5f-4c69-9a54-d1edaeeb01d0/submissions/77107902-c521-4482-b7f7-0da5b403888f/download'
stream = True, body = None
headers = {'Accept': 'application/json', 'Authorization': 'Token IkViNnlnaWFvczBCcFRMOHZUQnZkUFBoLVZYUF9XM255WVU2Q3h4SDVOUE0i.ZuN9ow.kmia5dehYOsb5MU_yBKi_DPZsbM', 'Content-Type': 'application/json', 'Range': 'bytes=650-'}
timeout = 3600

    def _send_json_request(
        self,
        method: str,
        path_query: str,
        stream: bool = False,
        body: str | None = None,
        headers: dict[str, str] | None = None,
        timeout: int | None = None,
    ) -> StreamedResponse | JSONResponse:
        """If the cassette contains a VCR.py `Request` object corresponding to
        this request, play back the response.  If it's an exception, raise it to
        be handled by the caller.
    
        Otherwise, make the request normally and record the response.  If it's
        an exception, record it anyway, then raise it to be handled by the
        caller.
        """
    
        request = Request(method, path_query, body, headers)
    
        # Playback mode:
        try:
>           response = self._cassette.play_response(request)

tests/sdk/utils.py:57: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <utils.Cassette object at 0x7b23c821b150>
request = <Request (GET) api/v1/sources/5b7d9dcf-df5f-4c69-9a54-d1edaeeb01d0/submissions/77107902-c521-4482-b7f7-0da5b403888f/download>

    def play_response(self, request):
        """
        Get the response corresponding to a request, but only if it
        hasn't been played back before, and mark it as played
        """
        for index, response in self._responses(request):
            if self.play_counts[index] == 0 or self.allow_playback_repeats:
                self.play_counts[index] += 1
                return response
        # The cassette doesn't contain the request asked for.
>       raise UnhandledHTTPRequestError(
            f"The cassette ({self._path!r}) doesn't contain the request ({request!r}) asked for",
        )
E       vcr.errors.UnhandledHTTPRequestError: "The cassette ('tests/sdk/data/test_download_submission_autoresume.yml') doesn't contain the request (<Request (GET) api/v1/sources/5b7d9dcf-df5f-4c69-9a54-d1edaeeb01d0/submissions/77107902-c521-4482-b7f7-0da5b403888f/download>) asked for"

../../securedrop/.venv/lib/python3.11/site-packages/vcr/cassette.py:262: UnhandledHTTPRequestError

During handling of the above exception, another exception occurred:

self = <test_api.TestAPI object at 0x7b23c85243d0>
mocker = <pytest_mock.plugin.MockerFixture object at 0x7b23c821ad90>

    @VCRAPI.use_cassette
    def test_download_submission_autoresume(self, mocker):
        # Get a submision
        submissions = self.api.get_all_submissions()
        for s in submissions:
            if s.is_file():
                submission = s
                break
    
        # Get the submission's content
        with tempfile.TemporaryDirectory() as tmpdir:
            _, filepath = self.api.download_submission(submission, tmpdir)
            with open(filepath, "rb") as fobj:
                content = fobj.read()
    
        # Stub subprocess.Popen to raise a subprocess.TimeoutExpired, to simulate an interrupted
        # download after 200 bytes
        class StubbedStdout:
            def __init__(self):
                self.counter = 0
                self.closed = False
    
            def read(self, n=-1):
                if self.counter >= 200:
                    # Restore the original Popen before raising the exception
                    mocker.stopall()
                    raise subprocess.TimeoutExpired(
                        cmd=None, timeout=None, output=None, stderr=None
                    )
                chunk = content[self.counter : self.counter + n]
                self.counter += len(chunk)
                return chunk
    
            def close(self):
                self.closed = True
    
            def fileno(self):
                return 1
    
        class StubbedStderr:
            def __init__(self):
                self.closed = False
    
            def read(self, n=-1):
                return json.dumps(
                    {
                        "headers": {
                            "content-type": "application/pgp-encrypted",
                            "etag": "sha256:3aa5ec3fe60b235a76bfc2c3a5c5d525687f04c1670060632a21b6a77c131f65",  # noqa: E501
                            "content-disposition": "attachment; filename=3-assessable_firmness-doc.gz.gpg",  # noqa: E501
                            "content-length": "651",
                            "accept-ranges": "bytes",
                        }
                    }
                ).encode()
    
            def close(self):
                self.closed = True
    
            def fileno(self):
                return 2
    
        class StubbedPopen(subprocess.Popen):
            def __init__(self, *args, content="", **kwargs):
                super().__init__(*args, **kwargs)
    
                # Only stub if the command contains "securedrop.Proxy" or "securedrop-proxy"
                cmd = shlex.join(args[0])
                if "securedrop.Proxy" in cmd or "securedrop-proxy" in cmd:
                    self._is_securedrop_proxy = True
                    self.stdout = StubbedStdout()
                    self.stderr = StubbedStderr()
                else:
                    self._is_securedrop_proxy = False
    
            def wait(self, timeout=None):
                if self._is_securedrop_proxy:
                    return 0
                else:
                    return super().wait()
    
        # Download again, with the stubbed Popen
        mocker.patch("subprocess.Popen", new=StubbedPopen)
        with tempfile.TemporaryDirectory() as tmpdir:
>           etag, filepath = self.api.download_submission(submission, tmpdir)

tests/sdk/test_api.py:280: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
securedrop_client/sdk/__init__.py:698: in download_submission
    response = self._send_json_request(
tests/sdk/utils.py:66: in _send_json_request
    response = super()._send_json_request(
securedrop_client/sdk/__init__.py:328: in _send_json_request
    return self._streaming_download(data, env)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <utils.VCRAPI object at 0x7b23c82182d0>
data = {'headers': {'Accept': 'application/json', 'Authorization': 'Token IkViNnlnaWFvczBCcFRMOHZUQnZkUFBoLVZYUF9XM255WVU2Q3h...s/5b7d9dcf-df5f-4c69-9a54-d1edaeeb01d0/submissions/77107902-c521-4482-b7f7-0da5b403888f/download', 'stream': True, ...}
env = {'SD_PROXY_ORIGIN': 'http://localhost:8081/'}

    def _streaming_download(
        self, data: dict[str, Any], env: dict
    ) -> StreamedResponse | JSONResponse:
        fobj = tempfile.TemporaryFile("w+b")
    
        retry = 0
        bytes_written = 0
        download_finished = False
    
        while not download_finished and retry < self.download_retry_limit:
            logger.debug(f"Streaming download, retry {retry}")
    
            try:
                # Update the range request if we're retrying
                if retry > 0:
                    data["headers"]["Range"] = f"bytes={bytes_written}-"
                    logger.debug(f"Retry {retry}, range: {bytes_written}-")
    
                # Serialize the data to send
                data_bytes = json.dumps(data).encode()
    
                # Open the process
                logger.debug(f"Retry {retry}, opening process")
                proc = subprocess.Popen(
                    self._rpc_target(),
                    stdin=subprocess.PIPE,
                    stdout=subprocess.PIPE,
                    stderr=subprocess.PIPE,
                    env=env,
                )
    
                # Send the input data
                if proc.stdin is not None:
                    logger.debug(f"Retry {retry}, sending data")
                    proc.stdin.write(data_bytes)
                    proc.stdin.close()
    
                # Write the contents to disk
                chunk_size = 1024
                while not download_finished:
                    if proc.stdout is not None:
                        chunk = proc.stdout.read(chunk_size)
                        if not chunk:
                            logger.debug(f"Retry {retry}, download finished")
                            download_finished = True
                            break
    
                    fobj.write(chunk)
                    bytes_written += len(chunk)
                    logger.debug(f"Retry {retry}, bytes written: {bytes_written:,}")
    
                # Wait for the process to end
                returncode = proc.wait()
                logger.debug(f"Retry {retry}, process ended with return code {returncode}")
    
                # Check for errors
                if returncode != 0:
                    # Actually, the download did not finish after all
                    download_finished = False
    
                    try:
                        if proc.stderr is not None:
                            error = json.loads(proc.stderr.read().decode())
                    except json.decoder.JSONDecodeError as err:
                        raise BaseError("Unable to parse stderr JSON") from err
                    raise BaseError("Internal proxy error: " + error.get("error", "unknown error"))
    
                # FIXME: For now, store the contents as bytes
                logger.debug(f"Retry {retry}, reading contents from disk")
                fobj.seek(0)
                contents = fobj.read()
                fobj.close()
    
                # Check for an error response
                if contents[0:1] == b"{":
                    logger.debug(f"Retry {retry}, received JSON error response")
                    return self._handle_json_response(contents)
    
                # Get the headers
                if proc.stderr is not None:
                    try:
                        stderr = json.loads(proc.stderr.read().decode())
                    except json.decoder.JSONDecodeError as err:
                        raise BaseError("Unable to parse stderr JSON") from err
    
                    sha256sum = stderr["headers"]["etag"]
                    filename = stderr["headers"]["content-disposition"]
                else:
                    # This should never happen because we should always have an stderr
                    sha256sum = ""
                    filename = ""
    
                return StreamedResponse(
                    contents=contents,
                    sha256sum=sha256sum,
                    filename=filename,
                )
    
            except subprocess.TimeoutExpired as err:
                logger.debug(f"Retry {retry}, timeout expired")
                retry += 1
                if retry >= self.download_retry_limit:
                    logger.debug("Retry limit reached, raising RequestTimeoutError")
                    raise RequestTimeoutError from err
            except BaseError as err:
                logger.debug(f"Retry {retry}, base error: {err}")
                retry += 1
                if retry >= self.download_retry_limit:
                    logger.debug("Retry limit reached, raising BaseError")
                    raise err
    
        # We will never reach this code because we'll have already returned or raised
        # an exception by now, but it's required to make the linter happy
>       raise RuntimeError(
            "This should be unreachable, we should've already returned or raised a different exception"  # noqa: E501
        )
E       RuntimeError: This should be unreachable, we should've already returned or raised a different exception

securedrop_client/sdk/__init__.py:265: RuntimeError
```


## Comments

My interpretation is that the current implementation of `StubbedStderr` assumes deterministic API results across test runs, which `make regenerate-sdk-cassettes` violates.  But see also #1992.

--- a/client/tests/sdk/test_api.py
+++ b/client/tests/sdk/test_api.py
@@ -233,7 +233,8 @@ class TestAPI(TestShared):
                 return 1
 
         class StubbedStderr:
-            def __init__(self):
+            def __init__(self, content_length):
+                self.content_length = content_length
                 self.closed = False
 
             def read(self, n=-1):
@@ -243,7 +244,7 @@ class TestAPI(TestShared):
                             "content-type": "application/pgp-encrypted",
                             "etag": "sha256:3aa5ec3fe60b235a76bfc2c3a5c5d525687f04c1670060632a21b6a77c131f65",  # noqa: E501
                             "content-disposition": "attachment; filename=3-assessable_firmness-doc.gz.gpg",  # noqa: E501
-                            "content-length": "651",
+                            "content-length": self.content_length,
                             "accept-ranges": "bytes",
                         }
                     }
@@ -264,7 +265,7 @@ class TestAPI(TestShared):
                 if "securedrop.Proxy" in cmd or "securedrop-proxy" in cmd:
                     self._is_securedrop_proxy = True
                     self.stdout = StubbedStdout()
-                    self.stderr = StubbedStderr()
+                    self.stderr = StubbedStderr(len(content))
                 else:
                     self._is_securedrop_proxy = False
 

